make-prs.py

```diff
@@ -0,0 +1,251 @@
+
+    if result.returncode != 0:
+        logger.error(f"Error pushing changes to branch {head}: {result.stderr}")
+        exit(1)
+
+
+def create_pr(repo: str, base: str, head: str, conflict: bool):
+    logger.info("Creating PR ...")
+
+    title = f"chore: merging diff into branch {base}"
+    body = f"Automated change: merging diff into branch {base}"
+
+    if conflict:
+        title = f"{title} CONFLICTS!"
+        body = f"**Conflicts! Need human intervention!**\n\n{body}"
+
+    repo = gh_client.get_repo(repo)
+    pr = repo.create_pull(base=base, head=head, title=title, body=body)
+    if not pr:
+        logger.error("Error creating PR.")
+        exit(1)
+    logger.info(f"PR {pr.number} created.")
+
+
+def get_all_chapters_branches(repo: str) -> typing.List[str]:
+    return sorted(
+        [
+            b.name
+            for b in gh_client.get_repo(repo).get_branches()
+            if re.search(r"\d", b.name)
+        ]
+    )
+
+
+def get_pr_base_branch(repo: str, pull_request_number: int) -> str:
+    return gh_client.get_repo(repo).get_pull(pull_request_number).base.ref
+
+
+def apply_diff_to_branch_and_create_pr(
+    repo: str, base: str, owner: str, ignore_conflicts: bool
+):
+    logger.info(f"=== Working on branch {base} ===")
+
+    head_branch = create_new_branch_based_on(base)
+
+    conflict = apply_patch()
+    if conflict and not ignore_conflicts:
+        return True
+
+    commit_diff(head_branch)
+
+    create_pr(repo, base, f"{owner}:{head_branch}", conflict)
+
+
+def cleanup():
+    logger.info("Cleaning up ...")
+
+    if os.path.exists(DIFF_FILE):
+        os.remove(DIFF_FILE)
+
+    shutil.rmtree(WORKING_DIR, ignore_errors=True)
+
+
+def main(
+    upstream_owner,
+    upstream_repo_name,
+    fork_owner,
+    fork_repo_name,
+    pull_request_number,
+    ignore_conflicts,
+):
+    upstream_repo = f"{upstream_owner}/{upstream_repo_name}"
+    fork_repo = f"{fork_owner}/{fork_repo_name}"
+
+    get_diff_as_patch(upstream_repo, pull_request_number)
+    clone(fork_repo)
+
+    pr_base = get_pr_base_branch(upstream_repo, pull_request_number)
+    target_branches = [
+        b for b in get_all_chapters_branches(upstream_repo) if b > pr_base
+    ]
+
+    for branch in target_branches:
+        conflict = apply_diff_to_branch_and_create_pr(
+            upstream_repo, branch, fork_owner, ignore_conflicts
+        )
+
+        if conflict and not ignore_conflicts:
+            logger.info(
+                f"Conflict merging diff into branch {branch}, aborting this PR and following PRs."
+            )
+            break
+
+    cleanup()  # comment this line out to keep tmp dir and diff file for debugging
+
+    logger.info("Done!")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        "-u",
+        "--upstream-owner",
+        type=str,
+        default="canonical",
+        help="The upstream owner of the repository, defaults to `canonical`.",
+    )
```

Comment:
WDYT about 3rd-party libs, like `typer` or `click`?

Given that we're using at least 2 other libraries, and a virtual environment is required, there isn't really much downside... I think.

Reply:
Let's leave this as Tiexin's written it: `argparse` might be a bit verbose, but it's not hard to read, and it's in the stdlib. I think it's okay to have a discussion about good 3rd party libs separately, but I don't think there's any need to rework what he's written here for this small script.

I'm a big fan of "use the stdlib if you can". There's not much downside in terms of using a lib given we already have a virtual env, but keeping 3rd party dependencies to a minimum is still a virtue (xz-style issues are much more likely to happen in `typer` or `click` than in Python itself). [Our Software Dependency Problem](https://research.swtch.com/deps) again. :-) Another reason: I (and I expect others) don't know either `typer` or `click`, but hopefully nearly everyone who uses Python knows the stdlib and `argparse`.